### PR TITLE
Maximum number of users

### PR DESCRIPTION
**Pull request by cjacobsen**
I need to limit maximum number of users on small Owncloud instances.

Its my first pull request, so please be patience :)

### CODE REVIEW

**Comment by MorrisJobke** [comment]
@owncloud-bot this is okay to test

**Comment by MorrisJobke** [comment]
cc @DeepDiver1975 @karlitschek

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
Thanks a lot @cjacobsen - I'm setting this to 8.2 as we are in the very last phase in the development cycle of 8.1.

File: settings/controller/userscontroller.php [hunk 1]
```diff
@@ -277,6 +277,17 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 	 * @return DataResponse
 	 */
 	public function create($username, $password, array $groups=array(), $email='') {
+		$sumUsers = array_sum($this->userManager->countUsers());
```
Comment (DeepDiver1975) [inline settings/controller/userscontroller.php]: This can we a pretty expensive call - thinking about being connected to LDAP

**Comment by DeepDiver1975** [comment]
From my understanding this is the wrong code place to implement this - there is also the console command which does not enter the userscontroller and therefore this limitation can be bypassed.

**Comment by DeepDiver1975** [comment]
> I need to limit maximum number of users on small Owncloud instances.

What is the use case for this feature? Users are created by the admin only - in our case of having a small instance I assume you are the only admin. I somewhat miss the necessity.

**Comment by DeepDiver1975** [comment]
> Its my first pull request, so please be patience :)

Welcome to the show! Thanks a lot for your contributions!

File: settings/controller/userscontroller.php [hunk 2]
```diff
@@ -277,6 +277,17 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
...
+		$maxUsers = $this->config->getSystemValue('max_users');
+		if (is_numeric($maxUsers) and $sumUsers >= $maxUsers) {
+			return new DataResponse(
+				array(
+					'message' => (string)$this->l10n->t('The maximum number of users is reached.').' ('.$sumUsers.'/'.$maxUsers.')'
```
Comment (nickvergessen) [inline settings/controller/userscontroller.php]: add the bracket information to the translation as well please

**Comment by cjacobsen** [comment]
The purpose is to prevent degradation of service in small installations, where the admin is not who creates all the users. The group admins have this responsability.

For example, if i create a instance with 2 group admins that should not create more than 20 users. 

In my case, i'm not using LDAP, but in this case, users are not created here, so it's not needed to check and the config variable should be called `max_local_users` or similar.

About the console command, it's not accesible by the group admin, but i'll check `occ` anyway.

Thanks for taking the time to review.

File: settings/controller/userscontroller.php [hunk 3]
```diff
@@ -277,6 +277,17 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 	 * @return DataResponse
 	 */
 	public function create($username, $password, array $groups=array(), $email='') {
+		$sumUsers = array_sum($this->userManager->countUsers());
+		$maxUsers = $this->config->getSystemValue('max_users');
+		if (is_numeric($maxUsers) and $sumUsers >= $maxUsers) {
+			return new DataResponse(
+				array(
+					'message' => (string)$this->l10n->t('The maximum number of users is reached. (%s/%s)', array($sumUsers, $maxUsers))
```
Comment (nickvergessen) [inline settings/controller/userscontroller.php]: Since those values should be numbers, you can use `%d` see http://php.net/manual/en/function.sprintf.php

**Comment by LukasReschke** [comment]
I'm not sure if this is something we should have. Technically admins can be do anything anyways and if this solely is about group admins creating users then we might want to have specific settings for subadmins instead of a config switch that nobody tests.

In your specific environment, @cjacobsen, I'd recommend to create an application that listens on `OC_User::pre_createUser` (old hook system) or `\OC\User::preCreateUser` (new hook system). It has a `&run` parameter that you can set to `false` to stop creating users.

Very simple example, put the following in an `appinfo/app.php`:

```php
\OCP\Util::connectHook('OC_User', 'pre_createUser', '\OCA\MyApp\MyClass', 'createUser');
```

And in `MyClass` do something like: 

```php
<?php

namespace \OCA\MyApp;

class MyClass {
	public function createUser(&$params) {
		// Stop running – that should suffice…
		$params['run'] = false;
                // FIXME: Huge hack around a bug…
		exit();
	}
}
```

@icewind1991 As per documentation `$params['run']` should be enough but with the new manager classes this seems to be not working any longer. Could you take a look? THX :smile:

**Comment by cjacobsen** [comment]
Thanks for the clarification. I'll follow your advice.

**Pull request closed without merging**